src/Servers/HttpSys/src/RequestProcessing/ResponseBody.cs

```diff
@@ -188,30 +192,41 @@ private unsafe void FlushInternal(bool endOfRequest, ArraySegment<byte> data = n
         }
     }
 
-    private List<GCHandle> PinDataBuffers(bool endOfRequest, ArraySegment<byte> data, out HttpApiTypes.HTTP_DATA_CHUNK[] dataChunks)
+    private unsafe Span<GCHandle> BuildDataChunks(scoped ref UnmanagedBufferAllocator allocator, bool endOfRequest, ArraySegment<byte> data, out Span<HttpApiTypes.HTTP_DATA_CHUNK> dataChunks)
     {
-        var pins = new List<GCHandle>();
+        const int maxGCHandleCount = 4;
+        int gcHandleIndex = 0;
+        var pins = allocator.AllocAsSpan<GCHandle>(maxGCHandleCount);
+
+        // Manually initialize the allocated GCHandles
+        for (int i = 0; i < maxGCHandleCount; ++i)
+        {
+            pins[i] = new GCHandle();
+        }
```

Comment:
It would make the method return `Span<HttpApiTypes.HTTP_DATA_CHUNK>`, pass `Span<GCHandle>` as out argument instead and assign the out argument here before we allocated any GCHandles.

It fixes the GCHandle memory leak in case there is an exception in the rest of this method.